pdudaemon/drivers/conrad197720.py

```diff
@@ -0,0 +1,325 @@
+        """
+        Private function to parse the amount of concatenated card at this serial port
+        :param self: The object itself
+        :return: number of cards
+        :rtype: int
+        """
+        self.num_cards = 0
+        for i in range(0, (len(data) >> 2)):
+            if (data[(i << 2)] == 254):
+                self.num_cards = self.num_cards + 1
+        log.debug(f"found {self.num_cards} cards")
+        return self.num_cards
+
+    def __getNumCards(self):
+        """
+        Return amount of cards found after init()
+        :param self: The object itself
+        :return: number of cards
+        :rtype: int
+        """
+        return self.num_cards
+
+    def __openConnection(self, portname):
+        """
+        Open serial connection
+        :param self: The object itself
+        :param portname: The string describing the serial device to open (e.g. "/dev/ttyUSB1" or "/dev/serial/by-id/...")
+        """
+        self.portname = portname
+        self.com.port = self.portname
+        self.com.baudrate = 19200
+        self.com.bytesize = 8
+        self.com.parity = serial.PARITY_NONE
+        self.com.stopbits = serial.STOPBITS_ONE
+        self.com.timeout = SERIAL_TIMEOUT
+        self.com.open()
+
+    def __init(self):
+        """
+        Init all cards
+        :param self: The object itself
+        :raises RuntimeError: all retries failed, communication error
+        """
+        self.__sendCommand(CMD_INIT, ADDR_FIRST_CARD, 0)
+        # make sure, all cards run with default setting
+        for card_addr in range(1, self.__getNumCards() + 1):
+            option = self.__sendCommand(CMD_GETOPTION, card_addr, 0)
+            if option != OPTION_BROADCAST_SEND_AND_NO_BLOCK:
+                # set card to default setting
+                log.debug(f"set option to default for card: {card_addr}")
+                self.__sendCommand(CMD_SETOPTION, card_addr, OPTION_BROADCAST_SEND_AND_NO_BLOCK)
+
+    def __updateSingle(self, command, port):
+        """
+        Update a single port of a card
+        :param self: The object itself
+        :param command: The command to execute, "on" "off" or "toggle"
+        :param port: The port number 0..n, all ports of all cards are in one range, e.g. port 9 is the second port of the second card
+        :return: True on success
+        :rtype: boolean
+        :raises RuntimeError: unknown command
+        :raises RuntimeError: no cards present
+        :raises RuntimeError: port number invalid
+        :raises RuntimeError: all retries failed, communication error
+        """
+        for i in range(0, RETRY_COUNT + 1):
+            card_addr = int(port / PORTS_PER_CARD) + 1
+            port_index = int(port % PORTS_PER_CARD)
+
+            if command == "off":
+                card_command = CMD_DELSINGLE
+            elif command == "on":
+                card_command = CMD_SETSINGLE
+            elif command == "toggle":
```

Comment:
the only valid value of command is on/off atm, so soggle will never be called

Reply:
my idea was to leave future relevant code in, in case pdudaemon supports toggle in future
but it's dead code, so i will remove it